### PR TITLE
adds a new export from the package for aws-sdk v3

### LINKED ISSUE #757: dynamic import bundles entire aws-sdk
### Is your feature request related to a problem?
When bundling opensearch-js with webpack or esbuild, the entire aws-sdk gets included in the bundle because of the import here https://github.com/opensearch-project/opensearch-js/blob/main/lib/aws/AwsSigv4Signer.js#L30C5-L30C43

This adds an additional 11MB to the final bundle size.

### What solution would you like?
The code can be refactored to make it a static import that can be treeshook, maybe separate files for aws sdk v2 & aws sdk v3.

### What alternatives have you considered?
We've marked aws-sdk as external to keep it from bundling but that's risky as some other library might need aws-sdk

### PR DESCRIPTION
**Pull request by tusharf5**
### Description

Based on the conversation in this issue #757 , this PR adds a new way to import the AwsSigv4Signer as shown below.  

```
const { AwsSigv4Signer } = require('@opensearch-project/opensearch/aws-v3');
```

By using a new import path, the sdk remains backwards compatible while also adding a way for clients to avoid getting the entire aws-sdk included in the final bundle which results in an additional 22MB.

I tested the code and it does reduce the bundle size with the new import path

image

The bundle size is still not good though because '@aws-sdk/credential-provider-node' dynamically imports all types of credential providers ini, ec2, environment, node, process, http etc which is about 1MB in size. 

Would it make sense to have the client provide the provider based on the environment instead of relying on the default provider chain? This would be a breaking change but it would delegate the credential responsibility on the client eliminating the dependency on aws-sdk & @aws-sdk/credential-provider-node?

https://github.com/opensearch-project/opensearch-js/blob/main/lib/aws/AwsSigv4Signer.js#L74

We will have to throw an error on this (making it required) line so the client always provides a `getCredential` function and we can remove the dependency on a default provider (both v2 and v3) from our end.

The doc examples already use the `getCredential` function so I am guessing most of the clients would already be providing this function and not relying on the default `getCredential`.

https://opensearch.org/docs/latest/clients/javascript/index/#authenticating-with-amazon-opensearch-service-aws-signature-version-4

### Issues Resolved

Closes #757 

### Check List

- [x] New functionality includes testing.
  - [x] All tests pass
- [x] Linter check was successfull - `yarn run lint` doesn't show any errors
- [x] Commits are signed per the DCO using --signoff
- [x] Changelog was updated.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: CHANGELOG.md
Comment (dblock) [inline CHANGELOG.md:7]: Match other CHANGELOG lines.

```
- Added `@opensearch-project/opensearch/aws-v3` import ([758](https://github.com/opensearch-project/opensearch-js/pull/758)).
```

File: lib/aws/AwsSigv4Signer-sdk-v3.js
Comment (dblock) [inline lib/aws/AwsSigv4Signer-sdk-v3.js:23]: This error is difficult to action. I think a simple "Missing AWS SDK" error is fine. If you want to offer a hint, try something that one can actually do (e.g. add an import? a dependency?).

**Review by dblock — CHANGES_REQUESTED** [review]
Good stuff. Document new functionality/changes in README, USERS_GUIDE, update samples. Add tests.

**Comment by dblock** [comment]
> Would it make sense to have the client provide the provider based on the environment instead of relying on the default provider chain? This would be a breaking change but it would delegate the credential responsibility on the client eliminating the dependency on aws-sdk & @aws-sdk/credential-provider-node?

It's a good idea, but I would explore this in a future/separate PR.

**Comment by tusharf5** [comment]
@dblock updated the PR with the requested changes

File: USER_GUIDE.md
Comment (dblock) [inline USER_GUIDE.md:93]: Should we say something about v2 here?

File: lib/aws/AwsSigv4Signer-sdk-v3.js
Comment (dblock) [inline lib/aws/AwsSigv4Signer-sdk-v3.js:23]: Add a period at the end of the sentence?

**Review by dblock — APPROVED** [review]
👏  This looks great to me!

Let's get another pair of eyes, maybe @nhtruong and merge?

**Comment by dblock** [comment]
@tusharf5 fix CI pls, there's some license headers missing, and a complaint about a new license, 0BSD. It's just a BSD variant so I think we can add it to the list, but I'll go double check.

File: USER_GUIDE.md
Comment (tusharf5) [reply USER_GUIDE.md:93]: I've added some more comments here

**Review by dblock — APPROVED** [review]

**Comment by dblock** [comment]
The license workflow needs to have 0BSD added to it, https://github.com/opensearch-project/opensearch-js/actions/runs/8739773216/job/23986128639?pr=758.

**Comment by tusharf5** [comment]
@dblock fixed

**Comment by tusharf5** [comment]
checking the test error

**Comment by tusharf5** [comment]
1 existing test was failing since aws-sdk and v3 ask were not installed so the error thrown was different and now they are part of dev dependency so the error has changed. I've updated the test. hopefully no more test failures. @dblock

File: test/unit/lib/aws/awssigv4signer.test.js
Comment (nhtruong) [inline test/unit/lib/aws/awssigv4signer.test.js:14]: Would it be easier and less confusing if we test this in a separate file/env?

Comment (nhtruong) [reply test/unit/lib/aws/awssigv4signer.test.js:14]: Probably 4 different envs:
- `AwsSigv4Signer` with aws-v2
- `AwsSigv4Signer` with aws-v3
- `AwsSigv4SignerV3` with aws-v3
- `AwsSigv4SignerV3` without aws-v3

Comment (tusharf5) [reply test/unit/lib/aws/awssigv4signer.test.js]: I've moved the tests to a different file

**Comment by tusharf5** [comment]
Had to remove the peerDependency change since npm automatically installs peerDepenencies (yarn does not) and that changes the import result which fails the test. It should be working now.

File: test/unit/lib/aws/awssignv4signer-versions.test.js
Comment (nhtruong) [inline test/unit/lib/aws/awssignv4signer-versions.test.js:91]: Are these tests only testing if they throw the correct errors not for paths where no errors are thrown? I think that's why Codecov is not happy.

Comment (tusharf5) [reply test/unit/lib/aws/awssignv4signer-versions.test.js:91]: This would require mocking the import as the dependencies are not installed. While tap supports mocking imports, the project does not use the latest tap version which has breaking changes. 

Doesn't seem like there is any other way to test dependency missing/present scenarios without mocking.

Comment (dblock) [reply test/unit/lib/aws/awssignv4signer-versions.test.js:91]: We should solve this with integration tests IMO, which is a separate affair.

Comment (nhtruong) [reply test/unit/lib/aws/awssignv4signer-versions.test.js:91]: I'm thinking of setting up different `package.json` for each test case like you have in [test/bundlers/esbuild-test/package.json](https://github.com/opensearch-project/opensearch-js/pull/758/files/b0eb72e76c8f17490dfb381664dd06d2cb9ec2fe#diff-1cb683c49678d4831e23c772e985f2d08b5af71d8c0d3cc931399df98768069c)

Comment (dblock) [reply test/unit/lib/aws/awssignv4signer-versions.test.js:91]: It's a good idea. Do we want to hold this PR until then? Open an issue and do it later?

**Pull request merged**
